packages/plugin-essentials/sources/commands/plugin/import/sources.ts

```diff
@@ -0,0 +1,99 @@
+import {tmpdir}                                                         from 'os';
+
+import {prepareRepo, runWorkflow}                                       from '../../set/version/sources';
+import {savePlugin}                                                     from '../import';
+
+import {getAvailablePlugins}                                            from '../list';
+
+const BUILD_WORKFLOW = ({pluginName, noMinify}: {noMinify: boolean, pluginName: string}, target: PortablePath) => [
+  [`yarn`, `build:${pluginName}`, ...noMinify ? [`--no-minify`] : [], `|`],
+];
+
+// eslint-disable-next-line arca/no-default-export
+export default class PluginDlSourcesCommand extends BaseCommand {
+  @Command.String()
+  name!: string;
+
+  @Command.String(`--path`)
+  installPath?: string;
+
+  @Command.String(`--repository`)
+  repository: string = `https://github.com/yarnpkg/berry.git`;
+
+  @Command.String(`--branch`)
+  branch: string = `master`;
+
+  @Command.Boolean(`--no-minify`)
+  noMinify: boolean = false;
+
+  @Command.Boolean(`-f,--force`)
+  force: boolean = false;
+
+  static usage: Usage = Command.Usage({
+    category: `Plugin-related commands`,
+    description: `build a plugin from sources`,
+    details: `
+      This command clones the Yarn repository into a temporary folder, builds the specified contrib plugin and updates the configuration to reference it in further CLI invocations.
+
+      The plugin can be referenced by name.
+    `,
+    examples: [[
+      `Build and activate the "@yarnpkg/plugin-exec" plugin`,
+      `$0 plugin import from sources @yarnpkg/plugin-exec`,
+    ], [
+      `Build and activate the "@yarnpkg/plugin-exec" plugin (shorthand)`,
+      `$0 plugin import from sources exec`,
+    ]],
+  });
+
+  @Command.Path(`plugin`, `import`, `from`, `sources`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+
+    const target = typeof this.installPath !== `undefined`
+      ? ppath.resolve(this.context.cwd, npath.toPortablePath(this.installPath))
+      : ppath.resolve(npath.toPortablePath(tmpdir()), `yarnpkg-sources` as Filename);
+
+    const report = await StreamReport.start({
+      configuration,
+      stdout: this.context.stdout,
+    }, async report => {
+      const {project} = await Project.find(configuration, this.context.cwd);
+
+      const ident = structUtils.parseIdent(this.name.replace(/^((@yarnpkg\/)?plugin-)?/, `@yarnpkg/plugin-`));
+      const identStr = structUtils.stringifyIdent(ident);
+      const data = await getAvailablePlugins(configuration);
+
+      if (!Object.prototype.hasOwnProperty.call(data, identStr))
+        throw new ReportError(MessageName.PLUGIN_NAME_NOT_FOUND, `Couldn't find a plugin named "${identStr}" on the remote registry. Note that only the plugins referenced on our website (https://github.com/yarnpkg/berry/blob/master/plugins.yml) can be built and imported from sources.`);
```

Comment:
That's not necessarily true, is it? Any plugin author can make their plugin installable through `from sources` if they follow the pattern we use (might be worth documenting this pattern in the command documentation, for good measure) 🤔

Reply:
This currently only works for our repo, because `getAvailablePlugins` and the `@yarnpkg/` scope are currently hardcoded. I'll look into what I need to do to make this work for third-party plugins.